### PR TITLE
Bug 1183923 - Front-end UI and publish to pulse for 'backfill' action

### PR DESCRIPTION
**Pull request by vaibhavmagarwal**
[Review on Reviewable](https://reviewable.io/reviews/mozilla/treeherder/760)

### CODE REVIEW

File: ui/plugins/pluginpanel.html [hunk 1]
```diff
@@ -107,6 +107,16 @@
               <span class="fa fa-bar-chart-o"></span>
             </a>
           </li>
+          <li>
+            <a title="Back fill this till last seen job"
```
Comment (tojonmz) [inline ui/plugins/pluginpanel.html]: Just a drive-by comment, it would be a bit clearer to use the words 'Backfill' and 'until' here.

File: ui/plugins/pluginpanel.html [hunk 2]
```diff
@@ -206,6 +206,13 @@
           <label>Log parsing status:</label>
           <span>No logs</span>
         </li>
+        <li class="small" ng-click="backfillJob()">
+          <label>
+            <a title="Trigger this jobtype back to the previous time it was run">
+                Backfill this job
+            </a>
+          </label>
+        </li>
```
Comment (tojonmz) [inline ui/plugins/pluginpanel.html]: The Job details panel is for passive actions (viewing, reading). If this is a flavor of retrigger imo I think it's more appropriate in the original location you had it (Job details action bar).

Comment (vaibhavmagarwal) [inline ui/plugins/pluginpanel.html]: @tojonmz thanks for your comment. :KWeirso commented ( https://bugzilla.mozilla.org/show_bug.cgi?id=1183923#c3 ) it to be not put in the "actions" bar since this will not be used as much. Thus I have shifted it to job details.

File: ui/plugins/controller.js [hunk 3]
```diff
@@ -314,6 +314,27 @@ treeherder.controller('PluginCtrl', [
...
+                    ThJobModel.backfill($scope.repoName, $scope.job.id).then(function() {
+                        thNotify.send("Initiated request for backfilling this job: " +
+                                       $scope.jobSearchStr, 'success');
+                    }).catch(function(e) {
+                        // Generic error eg. the user doesn't have LDAP access
+                        thNotify.send(
+                            ThModelErrors.format(e, "Unable to send backfill"), 'danger');
+                    });
+                } else {
+                    thNotify.send("Job not yet loaded for back-fill", 'warning');
```
Comment (tojonmz) [inline ui/plugins/controller.js]: `back-fill` should be `backfill` here.

File: ui/plugins/controller.js [hunk 4]
```diff
@@ -314,6 +314,27 @@ treeherder.controller('PluginCtrl', [
...
+                // Only backfill if we have a valid loaded job, if the user
+                // tries to backfill eg. via shortcut before the load we warn them
+                if ($scope.job.id) {
+                    ThJobModel.backfill($scope.repoName, $scope.job.id).then(function() {
+                        thNotify.send("Initiated request for backfilling this job: " +
+                                       $scope.jobSearchStr, 'success');
+                    }).catch(function(e) {
+                        // Generic error eg. the user doesn't have LDAP access
+                        thNotify.send(
+                            ThModelErrors.format(e, "Unable to send backfill"), 'danger');
```
Comment (tojonmz) [inline ui/plugins/controller.js]: Ed is pretty comfortable with 90-100chars line length, if you desired to move this up a line.

File: ui/plugins/pluginpanel.html (hunk 2 above)
Comment (tojonmz) [inline ui/plugins/pluginpanel.html]: I think we're at the point for creating a small nested drop down menu in the Job details action bar, either for general icon overflow or for new items like this where they don't need to be as prominent. We've an open bug for this but the number escapes me offhand. It's late here but I'll try to dig it up tomorrow.

File: ui/plugins/controller.js (hunk 4 above)
Comment (vaibhavmagarwal) [inline ui/plugins/controller.js]: If I bring the line up it goes to 101 chars :)
So lets leave it like this.

File: treeherder/model/derived/jobs.py [hunk 5]
```diff
@@ -326,6 +326,16 @@ def retrigger(self, requester, job):
         """
         self._job_action_event(job, 'retrigger', requester)
 
+    def backfill(self, requester, job):
+        """
+        Issue a retrigger to the given job
```
Comment (maurodoglio) [inline treeherder/model/derived/jobs.py]: I would rather use the same description you used for the backfill endpoint below.

File: ui/plugins/controller.js [hunk 6]
```diff
@@ -314,6 +314,27 @@ treeherder.controller('PluginCtrl', [
             }
         };
 
+        $scope.backfillJob = function() {
+            if ($scope.user.loggedin) {
+                // Only backfill if we have a valid loaded job, if the user
+                // tries to backfill eg. via shortcut before the load we warn them
+                if ($scope.job.id) {
+                    ThJobModel.backfill($scope.repoName, $scope.job.id).then(function() {
+                        thNotify.send("Initiated request for backfilling this job: " +
```
Comment (maurodoglio) [inline ui/plugins/controller.js]: That's too much info imho, this  is probably enough:
"Backfilling successfully requested"

File: ui/plugins/controller.js [hunk 7]
```diff
@@ -314,6 +314,27 @@ treeherder.controller('PluginCtrl', [
...
+                                       $scope.jobSearchStr, 'success');
+                    }).catch(function(e) {
```
Comment (maurodoglio) [inline ui/plugins/controller.js]: don't use catch here, you can pass this callback as second parameter to .then()

File: ui/js/models/job.js [hunk 8]
```diff
@@ -94,6 +94,17 @@ treeherder.factory('ThJobModel', [
         });
     };
 
+    ThJobModel.backfill = function(repoName, pk, config) {
+        config = config || {};
+        var timeout = config.timeout || null;
+
+        return $http.post(ThJobModel.get_uri(repoName)+pk+"/backfill/",
+            {timeout:timeout})
+            .then(function(response) {
```
Comment (maurodoglio) [inline ui/js/models/job.js]: You should delete this .then call because 1) the backfill endpoint doesn't return a job and 2) the code that calls this function doesn't expect it to return a ThJobModel.

File: ui/plugins/pluginpanel.html [hunk 9]
```diff
@@ -210,6 +210,13 @@
           <label>Taskcluster:</label>
           <span><a href="{{line.url}}" target="_blank">{{line.value}}</a></span>
         </li>
+        <li class="small" ng-click="backfillJob()" ng-show="user.is_staff">
+          <label>
+            <a title="Trigger this jobtype back to the previous time it was run">
```
Comment (maurodoglio) [inline ui/plugins/pluginpanel.html]: Given that this functionality is not available to all the users bot only to the sheriffs I would keep using the term backfill in the title as well. Or come up with a simpler description, like "Trigger this job on previous result-sets."

File: treeherder/webapp/api/jobs.py [hunk 10]
```diff
@@ -147,6 +147,20 @@ def retrigger(self, request, project, jm, pk=None):
         else:
             return Response("No job with id: {0}".format(pk), 404)
 
+    @action(permission_classes=[IsAuthenticated])
```
Comment (maurodoglio) [inline treeherder/webapp/api/jobs.py]: This permission class only check if a user is authenticated. As far as I understand you wanted to expose this functionality only to sheriffs, right? In order to do do that, you need to add a new permission class to treeherder/webapp/api/permissions.py, something like this:
```python
class IsStaff(permissions.BasePermission):

    """
    The request is authenticated as an admin staff (eg. sheriffs).
    """

    def has_permission(self, request, view):
        return (request.user and request.user.is_authenticated() and
                   request.user.is_staff)
```
and use that class to the action decorator

**Comment by maurodoglio** [comment]
Same as https://github.com/mozilla/treeherder/pull/747#issuecomment-121566147

File: treeherder/webapp/api/jobs.py (hunk 10 above)
Comment (vaibhavmagarwal) [inline treeherder/webapp/api/jobs.py]: There is already a class defined in permissions.py: https://github.com/mozilla/treeherder/blob/master/treeherder/webapp/api/permissions.py#L8
I will use that.

File: ui/plugins/pluginpanel.html (hunk 2 above)
Comment (tojonmz) [inline ui/plugins/pluginpanel.html]: > We've an open bug for this but the number escapes me offhand. It's late here but I'll try to dig it up tomorrow.

Well, we had something like it, but for jobs with multiple logs. Just fyi.
Bug [1113264](https://bugzilla.mozilla.org/show_bug.cgi?id=1113264)

Comment (vaibhavmagarwal) [inline ui/plugins/pluginpanel.html]: Thanks for that bug. Right now since this is a sheriff-only action, I guess sheriffs are ok to have it in the job details pane, we can move the "backfill" to action button when bug 1113264 is fixed.
